File: android/app/src/main/java/com/monsoonfire/portal/reference/PortalContracts.kt
package com.monsoonfire.portal.reference

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

// Canonical Kotlin mirror of web/src/api/portalContracts.ts

typealias PortalFnName = String
typealias PortalApiErrorCode = String

@Serializable
data class PortalApiErrorEnvelope(
    val ok: Boolean? = null,
    val error: String? = null,
    val message: String? = null,
    val code: PortalApiErrorCode? = null,
    val details: JsonElement? = null
)

@Serializable
data class PortalApiOkEnvelope(
    val ok: Boolean
)

@Serializable
data class PortalApiMeta(
    val atIso: String,
    val requestId: String,
    val fn: String,
    val url: String,
    val payload: JsonElement,
    val curlExample: String? = null,
    val status: Int? = null,
    val ok: Boolean? = null,
    val response: JsonElement? = null,
    val error: String? = null,
    val message: String? = null,
    val code: PortalApiErrorCode? = null
)

@Serializable
data class CreateBatchRequest(
    val ownerUid: String,
    val ownerDisplayName: String,
    val title: String,
    val kilnName: String? = null,
    val intakeMode: String,
    val estimatedCostCents: Int,
    val estimateNotes: String? = null,
    val notes: String? = null,
    val clientRequestId: String? = null
)

@Serializable
data class ReservationPreferredWindow(
    val earliestDate: String? = null,
    val latestDate: String? = null
)

@Serializable
data class CreateReservationRequest(
    val firingType: String,
    val shelfEquivalent: Double,
    val preferredWindow: ReservationPreferredWindow? = null,
    val linkedBatchId: String? = null
)

@Serializable
data class PickedUpAndCloseRequest(
    val uid: String,
    val batchId: String
)

@Serializable
data class ContinueJourneyRequest(
    val uid: String,
    val fromBatchId: String
)

@Serializable
data class UpdateReservationRequest(
    val reservationId: String,
    val status: String,
    val staffNotes: String? = null
)

@Serializable
data class AssignReservationStationRequest(
    val reservationId: String,
    val assignedStationId: String,
    val queueClass: String? = null,
    val requiredResources: AssignedStationRequiredResources? = null
)

@Serializable
data class AssignedStationRequiredResources(
    val kilnProfile: String? = null,
    val rackCount: Int? = null,
    val specialHandling: List<String>? = null
)

@Serializable
data class MaterialsCartItemRequest(
    val productId: String,
    val quantity: Int
)

@Serializable
data class ListMaterialsProductsRequest(
    val includeInactive: Boolean? = null
)

@Serializable
data class CreateMaterialsCheckoutSessionRequest(
    val items: List<MaterialsCartItemRequest>,
    val pickupNotes: String? = null
)

@Serializable
data class SeedMaterialsCatalogRequest(
    val force: Boolean? = null
)

@Serializable
data class ListEventsRequest(
    val includeDrafts: Boolean? = null,
    val includeCancelled: Boolean? = null
)

@Serializable
data class GetEventRequest(
    val eventId: String
)

@Serializable
data class ListEventSignupsRequest(
    val eventId: String,
    val includeCancelled: Boolean? = null,
    val includeExpired: Boolean? = null,
    val limit: Int? = null
)

@Serializable
data class ListBillingSummaryRequest(
    val limit: Int? = null,
    val from: String? = null,
    val to: String? = null
)

@Serializable
data class ImportLibraryIsbnsRequest(
    val isbns: List<String>,
    val source: String? = null
)

@Serializable
data class RegisterDeviceTokenRequest(
    val token: String,
    val platform: String? = null,
    val environment: String? = null,
    val appVersion: String? = null,
    val appBuild: String? = null,
    val deviceModel: String? = null
)

@Serializable
data class UnregisterDeviceTokenRequest(
    val token: String? = null,
    val tokenHash: String? = null
)

@Serializable
data class RunNotificationFailureDrillRequest(
    val uid: String,
    val mode: String,
    val channels: NotificationFailureDrillChannels? = null,
    val forceRunNow: Boolean? = null
)

@Serializable
data class NotificationFailureDrillChannels(
    val inApp: Boolean? = null,
    val email: Boolean? = null,
    val push: Boolean? = null
)

@Serializable
data class RunNotificationMetricsAggregationNowRequest(
)

@Serializable
data class EventAddOnInput(
    val id: String,
    val title: String,
    val priceCents: Int,
    val isActive: Boolean
)

@Serializable
data class CreateEventRequest(
    val templateId: String? = null,
    val title: String,
    val summary: String,
    val description: String,
    val location: String,
    val timezone: String,
    val startAt: String,
    val endAt: String,
    val capacity: Int,
    val priceCents: Int,
    val currency: String,
    val includesFiring: Boolean,
    val firingDetails: String? = null,
    val policyCopy: String? = null,
    val addOns: List<EventAddOnInput>? = null,
    val waitlistEnabled: Boolean? = null,
    val offerClaimWindowHours: Int? = null,
    val cancelCutoffHours: Int? = null
)

@Serializable
data class PublishEventRequest(
    val eventId: String
)

@Serializable
data class SignupForEventRequest(
    val eventId: String
)

@Serializable
data class CancelEventSignupRequest(
    val signupId: String
)

@Serializable
data class ClaimEventOfferRequest(
    val signupId: String
)

@Serializable
data class CheckInEventRequest(
    val signupId: String,
    val method: String
)

@Serializable
data class CreateEventCheckoutSessionRequest(
    val eventId: String,
    val signupId: String,
    val addOnIds: List<String>? = null
)

@Serializable
data class CreateBatchResponse(
    val ok: Boolean,
    val batchId: String? = null,
    val newBatchId: String? = null,
    val existingBatchId: String? = null
)

@Serializable
data class PickedUpAndCloseResponse(
    val ok: Boolean
)

@Serializable
data class ContinueJourneyResponse(
    val ok: Boolean,
    val batchId: String? = null,
    val newBatchId: String? = null,
    val existingBatchId: String? = null,
    val rootId: String? = null,
    val fromBatchId: String? = null,
    val message: String? = null
)

@Serializable
data class CreateReservationResponse(
    val ok: Boolean,
    val reservationId: String? = null,
    val status: String? = null
)

@Serializable
data class UpdateReservationResponse(
    val ok: Boolean,
    val reservationId: String? = null,
    val status: String? = null
)

@Serializable
data class AssignReservationStationResponse(
    val ok: Boolean,
    val reservationId: String? = null,
    val assignedStationId: String? = null,
    val previousAssignedStationId: String? = null,
    val stationCapacity: Int? = null,
    val stationUsedAfter: Int? = null,
    val idempotentReplay: Boolean? = null
)

typealias EventStatus = String
typealias EventSignupStatus = String
typealias EventPaymentStatus = String

@Serializable
data class MaterialProduct(
    val id: String,
    val name: String,
    val description: String? = null,
    val category: String? = null,
    val sku: String? = null,
    val priceCents: Int,
    val currency: String,
    val stripePriceId: String? = null,
    val imageUrl: String? = null,
    val trackInventory: Boolean,
    val inventoryOnHand: Int? = null,
    val inventoryReserved: Int? = null,
    val inventoryAvailable: Int? = null,
    val active: Boolean
)

@Serializable
data class ListMaterialsProductsResponse(
    val ok: Boolean,
    val products: List<MaterialProduct>
)

@Serializable
data class CreateMaterialsCheckoutSessionResponse(
    val ok: Boolean,
    val orderId: String,
    val checkoutUrl: String? = null
)

@Serializable
data class SeedMaterialsCatalogResponse(
    val ok: Boolean,
    val created: Int,
    val updated: Int,
    val total: Int
)

@Serializable
data class EventAddOn(
    val id: String,
    val title: String,
    val priceCents: Int,
    val isActive: Boolean
)

@Serializable
data class EventSummary(
    val id: String,
    val title: String,
    val summary: String,
    val startAt: String? = null,
    val endAt: String? = null,
    val timezone: String? = null,
    val location: String? = null,
    val priceCents: Int,
    val currency: String,
    val includesFiring: Boolean,
    val firingDetails: String? = null,
    val capacity: Int,
    val waitlistEnabled: Boolean,
    val status: String,
    val remainingCapacity: Int? = null
)

@Serializable
data class EventDetail(
    val id: String,
    val title: String,
    val summary: String,
    val description: String,
    val startAt: String? = null,
    val endAt: String? = null,
    val timezone: String? = null,
    val location: String? = null,
    val priceCents: Int,
    val currency: String,
    val includesFiring: Boolean,
    val firingDetails: String? = null,
    val policyCopy: String? = null,
    val addOns: List<EventAddOn>? = null,
    val capacity: Int,
    val waitlistEnabled: Boolean,
    val offerClaimWindowHours: Int? = null,
    val cancelCutoffHours: Int? = null,
    val status: String
)

@Serializable
data class EventSignupSummary(
    val id: String,
    val status: String,
    val paymentStatus: String? = null
)

@Serializable
data class EventSignupRosterEntry(
    val id: String,
    val uid: String? = null,
    val displayName: String? = null,
    val email: String? = null,
    val status: String,
    val paymentStatus: String? = null,
    val createdAt: String? = null,
    val offerExpiresAt: String? = null,
    val checkedInAt: String? = null,
    val checkInMethod: String? = null
)

@Serializable
data class ListEventsResponse(
    val ok: Boolean,
    val events: List<EventSummary>
)

@Serializable
data class ListEventSignupsResponse(
    val ok: Boolean,
    val signups: List<EventSignupRosterEntry>
)

@Serializable
data class GetEventResponse(
    val ok: Boolean,
    val event: EventDetail,
    val signup: EventSignupSummary? = null
)

@Serializable
data class CreateEventResponse(
    val ok: Boolean,
    val eventId: String
)

@Serializable
data class PublishEventResponse(
    val ok: Boolean,
    val status: String
)

@Serializable
data class SignupForEventResponse(
    val ok: Boolean,
    val signupId: String,
    val status: String
)

@Serializable
data class CancelEventSignupResponse(
    val ok: Boolean,
    val status: String
)

@Serializable
data class ClaimEventOfferResponse(
    val ok: Boolean,
    val status: String
)

@Serializable
data class CheckInEventResponse(
    val ok: Boolean,
    val status: String,
    val paymentStatus: String? = null
)

@Serializable
data class CreateEventCheckoutSessionResponse(
    val ok: Boolean,
    val checkoutUrl: String? = null
)

@Serializable
data class ImportLibraryIsbnError(
    val isbn: String,
    val message: String
)

@Serializable
data class ImportLibraryIsbnsResponse(
    val ok: Boolean,
    val requested: Int,
    val created: Int,
    val updated: Int,
    val errors: List<ImportLibraryIsbnError>? = null
)

@Serializable
data class RegisterDeviceTokenResponse(
    val ok: Boolean,
    val uid: String,
    val tokenHash: String
)

@Serializable
data class UnregisterDeviceTokenResponse(
    val ok: Boolean,
    val uid: String,
    val tokenHash: String
)

@Serializable
data class RunNotificationFailureDrillResponse(
    val ok: Boolean,
    val jobId: String,
    val uid: String,
    val mode: String
)

@Serializable
data class RunNotificationMetricsAggregationNowResponse(
    val ok: Boolean,
    val windowHours: Int,
    val totalAttempts: Int,
    val statusCounts: Map<String, Int>,
    val reasonCounts: Map<String, Int>,
    val providerCounts: Map<String, Int>
)

@Serializable
data class MaterialOrderItemSummary(
    val productId: String,
    val name: String,
    val quantity: Int,
    val unitPrice: Int,
    val currency: String
)

@Serializable
data class MaterialOrderSummary(
    val id: String,
    val status: String,
    val totalCents: Int,
    val currency: String,
    val pickupNotes: String? = null,
    val checkoutUrl: String? = null,
    val createdAt: String? = null,
    val updatedAt: String? = null,
    val items: List<MaterialOrderItemSummary> = emptyList()
)

@Serializable
data class BillingReceipt(
    val id: String,
    val type: String,
    val sourceId: String? = null,
    val title: String,
    val amountCents: Int,
    val currency: String,
    val paidAt: String? = null,
    val createdAt: String? = null,
    val metadata: JsonElement? = null
)

@Serializable
data class BillingSummaryTotals(
    val unpaidCheckInsCount: Int,
    val unpaidCheckInsAmountCents: Int,
    val materialsPendingCount: Int,
    val materialsPendingAmountCents: Int,
    val receiptsCount: Int,
    val receiptsAmountCents: Int
)

@Serializable
data class BillingSummaryResponse(
    val ok: Boolean,
    val unpaidCheckIns: List<EventSignupRosterEntry> = emptyList(),
    val materialsOrders: List<MaterialOrderSummary> = emptyList(),
    val receipts: List<BillingReceipt> = emptyList(),
    val summary: BillingSummaryTotals
)

fun getResultBatchId(resp: CreateBatchResponse?): String? {
    if (resp == null) return null
    return resp.newBatchId ?: resp.batchId ?: resp.existingBatchId
}

@Serializable
enum class TimelineEventType {
    @SerialName("CREATE_BATCH")
    CREATE_BATCH,
    @SerialName("SUBMIT_DRAFT")
    SUBMIT_DRAFT,
    @SerialName("SHELVED")
    SHELVED,
    @SerialName("KILN_LOAD")
    KILN_LOAD,
    @SerialName("KILN_UNLOAD")
    KILN_UNLOAD,
    @SerialName("ASSIGNED_FIRING")
    ASSIGNED_FIRING,
    @SerialName("READY_FOR_PICKUP")
    READY_FOR_PICKUP,
    @SerialName("PICKED_UP_AND_CLOSE")
    PICKED_UP_AND_CLOSE,
    @SerialName("CONTINUE_JOURNEY")
    CONTINUE_JOURNEY,
}

val TIMELINE_EVENT_LABELS: Map<TimelineEventType, String> = mapOf(
    TimelineEventType.CREATE_BATCH to "Batch created",
    TimelineEventType.SUBMIT_DRAFT to "Draft submitted",
    TimelineEventType.SHELVED to "Shelved",
    TimelineEventType.KILN_LOAD to "Loaded into kiln",
    TimelineEventType.KILN_UNLOAD to "Unloaded from kiln",
    TimelineEventType.ASSIGNED_FIRING to "Firing assigned",
    TimelineEventType.READY_FOR_PICKUP to "Ready for pickup",
    TimelineEventType.PICKED_UP_AND_CLOSE to "Picked up & closed",
    TimelineEventType.CONTINUE_JOURNEY to "Journey continued",
)

fun normalizeTimelineEventType(raw: String?): TimelineEventType? {
    if (raw == null) return null
    return when (raw) {
        "CREATE_BATCH" -> TimelineEventType.CREATE_BATCH
        "SUBMIT_DRAFT" -> TimelineEventType.SUBMIT_DRAFT
        "SHELVED" -> TimelineEventType.SHELVED
        "KILN_LOAD" -> TimelineEventType.KILN_LOAD
        "KILN_UNLOAD" -> TimelineEventType.KILN_UNLOAD
        "ASSIGNED_FIRING" -> TimelineEventType.ASSIGNED_FIRING
        "READY_FOR_PICKUP" -> TimelineEventType.READY_FOR_PICKUP
        "PICKED_UP_AND_CLOSE" -> TimelineEventType.PICKED_UP_AND_CLOSE
        "CONTINUE_JOURNEY" -> TimelineEventType.CONTINUE_JOURNEY
        "BATCH_CREATED" -> TimelineEventType.CREATE_BATCH
        "SUBMITTED" -> TimelineEventType.SUBMIT_DRAFT
        "PICKED_UP_AND_CLOSED" -> TimelineEventType.PICKED_UP_AND_CLOSE
        else -> null
    }
}

File: ios/PortalContracts.swift
import Foundation

// MARK: - Portal API Contracts
// Canonical Swift mirror of web/src/api/portalContracts.ts

typealias PortalFnName = String
typealias PortalApiErrorCode = String

struct PortalApiErrorEnvelope: Decodable {
    var ok: Bool?
    var error: String?
    var message: String?
    var code: PortalApiErrorCode?
    var details: JSONValue?
}

struct PortalApiOkEnvelope: Decodable {
    var ok: Bool
}

// MARK: - Troubleshooting meta (matches PortalApiMeta in TS)

struct PortalApiMeta: Codable {
    var atIso: String
    var requestId: String
    var fn: String
    var url: String

    var payload: JSONValue
    var curlExample: String?

    var status: Int?
    var ok: Bool?

    var response: JSONValue?

    var error: String?
    var message: String?
    var code: PortalApiErrorCode?
}

// MARK: - JSONValue (for storing unknown JSON in meta)

enum JSONValue: Codable {
    case null
    case bool(Bool)
    case number(Double)
    case string(String)
    case array([JSONValue])
    case object([String: JSONValue])

    init(from decoder: Decoder) throws {
        let c = try decoder.singleValueContainer()
        if c.decodeNil() { self = .null; return }
        if let b = try? c.decode(Bool.self) { self = .bool(b); return }
        if let n = try? c.decode(Double.self) { self = .number(n); return }
        if let s = try? c.decode(String.self) { self = .string(s); return }
        if let a = try? c.decode([JSONValue].self) { self = .array(a); return }
        if let o = try? c.decode([String: JSONValue].self) { self = .object(o); return }
        throw DecodingError.dataCorruptedError(in: c, debugDescription: "Unknown JSONValue")
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.singleValueContainer()
        switch self {
        case .null: try c.encodeNil()
        case .bool(let b): try c.encode(b)
        case .number(let n): try c.encode(n)
        case .string(let s): try c.encode(s)
        case .array(let a): try c.encode(a)
        case .object(let o): try c.encode(o)
        }
    }
}

// MARK: - Requests

struct CreateBatchRequest: Codable {
    let ownerUid: String
    let ownerDisplayName: String
    let title: String
    let kilnName: String?
    let intakeMode: String
    let estimatedCostCents: Int
    let estimateNotes: String?
    let notes: String?
    let clientRequestId: String?
}

struct ReservationPreferredWindow: Codable {
    let earliestDate: String?
    let latestDate: String?
}

struct CreateReservationRequest: Codable {
    let firingType: String
    let shelfEquivalent: Double
    let footprintHalfShelves: Double?
    let heightInches: Double?
    let tiers: Int?
    let estimatedHalfShelves: Double?
    let useVolumePricing: Bool?
    let volumeIn3: Double?
    let estimatedCost: Double?
    let preferredWindow: ReservationPreferredWindow?
    let linkedBatchId: String?
    let clientRequestId: String?
    let ownerUid: String?
    let wareType: String?
    let kilnId: String?
    let kilnLabel: String?
    let quantityTier: String?
    let quantityLabel: String?
    let photoUrl: String?
    let photoPath: String?
    let dropOffProfile: ReservationDropOffProfile?
    let dropOffQuantity: ReservationDropOffQuantity?
    let notes: ReservationNotes?
    let addOns: ReservationAddOns?
}

struct ReservationDropOffProfile: Codable {
    let id: String?
    let label: String?
    let pieceCount: String?
    let hasTall: Bool?
    let stackable: Bool?
    let bisqueOnly: Bool?
    let specialHandling: Bool?
}

struct ReservationDropOffQuantity: Codable {
    let id: String?
    let label: String?
    let pieceRange: String?
}

struct ReservationNotes: Codable {
    let general: String?
    let clayBody: String?
    let glazeNotes: String?
}

struct ReservationAddOns: Codable {
    let rushRequested: Bool?
    let wholeKilnRequested: Bool?
    let pickupDeliveryRequested: Bool?
    let returnDeliveryRequested: Bool?
    let useStudioGlazes: Bool?
    let glazeAccessCost: Double?
}

struct PickedUpAndCloseRequest: Codable {
    let uid: String
    let batchId: String
}

struct ContinueJourneyRequest: Codable {
    let uid: String
    let fromBatchId: String
}

struct UpdateReservationRequest: Codable {
    let reservationId: String
    let status: String
    let staffNotes: String?
}

struct AssignReservationStationRequest: Codable {
    let reservationId: String
    let assignedStationId: String
    let queueClass: String?
    let requiredResources: ReservationStationResources?
}

struct ReservationStationResources: Codable {
    let kilnProfile: String?
    let rackCount: Int?
    let specialHandling: [String]?
}

struct MaterialsCartItemRequest: Codable {
    let productId: String
    let quantity: Int
}

struct ListMaterialsProductsRequest: Codable {
    let includeInactive: Bool?
}

struct CreateMaterialsCheckoutSessionRequest: Codable {
    let items: [MaterialsCartItemRequest]
    let pickupNotes: String?
}

struct SeedMaterialsCatalogRequest: Codable {
    let force: Bool?
}

struct ListEventsRequest: Codable {
    let includeDrafts: Bool?
    let includeCancelled: Bool?
}

struct GetEventRequest: Codable {
    let eventId: String
}

struct ListEventSignupsRequest: Codable {
    let eventId: String
    let includeCancelled: Bool?
    let includeExpired: Bool?
    let limit: Int?
}

struct ListBillingSummaryRequest: Codable {
    let limit: Int?
    let from: String?
    let to: String?
}

struct EventAddOnInput: Codable {
    let id: String
    let title: String
    let priceCents: Int
    let isActive: Bool
}

struct CreateEventRequest: Codable {
    let templateId: String?
    let title: String
    let summary: String
    let description: String
    let location: String
    let timezone: String
    let startAt: String
    let endAt: String
    let capacity: Int
    let priceCents: Int
    let currency: String
    let includesFiring: Bool
    let firingDetails: String?
    let policyCopy: String?
    let addOns: [EventAddOnInput]?
    let waitlistEnabled: Bool?
    let offerClaimWindowHours: Int?
    let cancelCutoffHours: Int?
}

struct PublishEventRequest: Codable {
    let eventId: String
}

struct SignupForEventRequest: Codable {
    let eventId: String
}

struct CancelEventSignupRequest: Codable {
    let signupId: String
}

struct ClaimEventOfferRequest: Codable {
    let signupId: String
}

struct CheckInEventRequest: Codable {
    let signupId: String
    let method: String
}

struct CreateEventCheckoutSessionRequest: Codable {
    let eventId: String
    let signupId: String
    let addOnIds: [String]?
}

struct ImportLibraryIsbnsRequest: Codable {
    let isbns: [String]
    let source: String?
}

struct RegisterDeviceTokenRequest: Codable {
    let token: String
    let platform: String?
    let environment: String?
    let appVersion: String?
    let appBuild: String?
    let deviceModel: String?
}

struct UnregisterDeviceTokenRequest: Codable {
    let token: String?
    let tokenHash: String?
}

struct RunNotificationFailureDrillRequest: Codable {
    let uid: String
    let mode: String
    let channels: DrillChannels?
    let forceRunNow: Bool?
}

struct DrillChannels: Codable {
    let inApp: Bool?
    let email: Bool?
    let push: Bool?
}

struct RunNotificationMetricsAggregationNowRequest: Codable {}

// MARK: - Responses

struct CreateBatchResponse: Codable {
    let ok: Bool
    let batchId: String?
    let newBatchId: String?
    let existingBatchId: String?
}

struct PickedUpAndCloseResponse: Codable {
    let ok: Bool
}

struct ContinueJourneyResponse: Codable {
    let ok: Bool
    let batchId: String?
    let newBatchId: String?
    let existingBatchId: String?
    let rootId: String?
    let fromBatchId: String?
    let message: String?
}

struct CreateReservationResponse: Codable {
    let ok: Bool
    let reservationId: String?
    let status: String?
}

struct UpdateReservationResponse: Codable {
    let ok: Bool
    let reservationId: String?
    let status: String?
}

struct AssignReservationStationResponse: Codable {
    let ok: Bool
    let reservationId: String?
    let assignedStationId: String?
    let previousAssignedStationId: String?
    let stationCapacity: Int?
    let stationUsedAfter: Int?
    let idempotentReplay: Bool?
}

struct MaterialProduct: Codable {
    let id: String
    let name: String
    let description: String?
    let category: String?
    let sku: String?
    let priceCents: Int
    let currency: String
    let stripePriceId: String?
    let imageUrl: String?
    let trackInventory: Bool
    let inventoryOnHand: Int?
    let inventoryReserved: Int?
    let inventoryAvailable: Int?
    let active: Bool
}

struct ListMaterialsProductsResponse: Codable {
    let ok: Bool
    let products: [MaterialProduct]
}

struct CreateMaterialsCheckoutSessionResponse: Codable {
    let ok: Bool
    let orderId: String
    let checkoutUrl: String?
}

struct SeedMaterialsCatalogResponse: Codable {
    let ok: Bool
    let created: Int
    let updated: Int
    let total: Int
}

struct EventAddOn: Codable {
    let id: String
    let title: String
    let priceCents: Int
    let isActive: Bool
}

typealias EventStatus = String
typealias EventSignupStatus = String
typealias EventPaymentStatus = String

struct EventSummary: Codable {
    let id: String
    let title: String
    let summary: String
    let startAt: String?
    let endAt: String?
    let timezone: String?
    let location: String?
    let priceCents: Int
    let currency: String
    let includesFiring: Bool
    let firingDetails: String?
    let capacity: Int
    let waitlistEnabled: Bool
    let status: String
    let remainingCapacity: Int?
}

struct EventDetail: Codable {
    let id: String
    let title: String
    let summary: String
    let description: String
    let startAt: String?
    let endAt: String?
    let timezone: String?
    let location: String?
    let priceCents: Int
    let currency: String
    let includesFiring: Bool
    let firingDetails: String?
    let policyCopy: String?
    let addOns: [EventAddOn]?
    let capacity: Int
    let waitlistEnabled: Bool
    let offerClaimWindowHours: Int?
    let cancelCutoffHours: Int?
    let status: String
}

struct EventSignupSummary: Codable {
    let id: String
    let status: String
    let paymentStatus: String?
}

struct EventSignupRosterEntry: Codable {
    let id: String
    let uid: String?
    let displayName: String?
    let email: String?
    let status: String
    let paymentStatus: String?
    let createdAt: String?
    let offerExpiresAt: String?
    let checkedInAt: String?
    let checkInMethod: String?
}

struct ListEventsResponse: Codable {
    let ok: Bool
    let events: [EventSummary]
}

struct ListEventSignupsResponse: Codable {
    let ok: Bool
    let signups: [EventSignupRosterEntry]
}

struct GetEventResponse: Codable {
    let ok: Bool
    let event: EventDetail
    let signup: EventSignupSummary?
}

struct CreateEventResponse: Codable {
    let ok: Bool
    let eventId: String
}

struct PublishEventResponse: Codable {
    let ok: Bool
    let status: String
}

struct SignupForEventResponse: Codable {
    let ok: Bool
    let signupId: String
    let status: String
}

struct CancelEventSignupResponse: Codable {
    let ok: Bool
    let status: String
}

struct ClaimEventOfferResponse: Codable {
    let ok: Bool
    let status: String
}

struct CheckInEventResponse: Codable {
    let ok: Bool
    let status: String
    let paymentStatus: String?
}

struct CreateEventCheckoutSessionResponse: Codable {
    let ok: Bool
    let checkoutUrl: String?
}

struct ImportLibraryIsbnError: Codable {
    let isbn: String
    let message: String
}

struct ImportLibraryIsbnsResponse: Codable {
    let ok: Bool
    let requested: Int
    let created: Int
    let updated: Int
    let errors: [ImportLibraryIsbnError]?
}

struct RegisterDeviceTokenResponse: Codable {
    let ok: Bool
    let uid: String
    let tokenHash: String
}

struct UnregisterDeviceTokenResponse: Codable {
    let ok: Bool
    let uid: String
    let tokenHash: String
}

struct RunNotificationFailureDrillResponse: Codable {
    let ok: Bool
    let jobId: String
    let uid: String
    let mode: String
}

struct RunNotificationMetricsAggregationNowResponse: Codable {
    let ok: Bool
    let windowHours: Int
    let totalAttempts: Int
    let statusCounts: [String: Int]
    let reasonCounts: [String: Int]
    let providerCounts: [String: Int]
}

struct MaterialOrderItemSummary: Codable {
    let productId: String
    let name: String
    let quantity: Int
    let unitPrice: Int
    let currency: String
}

struct MaterialOrderSummary: Codable {
    let id: String
    let status: String
    let totalCents: Int
    let currency: String
    let pickupNotes: String?
    let checkoutUrl: String?
    let createdAt: String?
    let updatedAt: String?
    let items: [MaterialOrderItemSummary]
}

enum BillingReceiptType: String, Codable {
    case event
    case materials
}

struct BillingReceipt: Codable {
    let id: String
    let type: BillingReceiptType
    let sourceId: String?
    let title: String
    let amountCents: Int
    let currency: String
    let paidAt: String?
    let createdAt: String?
    let metadata: [String: JSONValue]?
}

struct BillingSummaryTotals: Codable {
    let unpaidCheckInsCount: Int
    let unpaidCheckInsAmountCents: Int
    let materialsPendingCount: Int
    let materialsPendingAmountCents: Int
    let receiptsCount: Int
    let receiptsAmountCents: Int
}

struct BillingSummaryResponse: Codable {
    let ok: Bool
    let unpaidCheckIns: [EventSignupRosterEntry]
    let materialsOrders: [MaterialOrderSummary]
    let receipts: [BillingReceipt]
    let summary: BillingSummaryTotals
}

func getResultBatchId(
    _ resp: CreateBatchResponse?
) -> String? {
    guard let resp else { return nil }
    return resp.newBatchId ?? resp.batchId ?? resp.existingBatchId
}

// MARK: - Timeline events (shared)

enum TimelineEventType: String, Codable {
    case createBatch = "CREATE_BATCH"
    case submitDraft = "SUBMIT_DRAFT"
    case shelved = "SHELVED"
    case kilnLoad = "KILN_LOAD"
    case kilnUnload = "KILN_UNLOAD"
    case assignedFiring = "ASSIGNED_FIRING"
    case readyForPickup = "READY_FOR_PICKUP"
    case pickedUpAndClose = "PICKED_UP_AND_CLOSE"
    case continueJourney = "CONTINUE_JOURNEY"
}

let TIMELINE_EVENT_LABELS: [TimelineEventType: String] = [
    .createBatch: "Batch created",
    .submitDraft: "Draft submitted",
    .shelved: "Shelved",
    .kilnLoad: "Loaded into kiln",
    .kilnUnload: "Unloaded from kiln",
    .assignedFiring: "Firing assigned",
    .readyForPickup: "Ready for pickup",
    .pickedUpAndClose: "Picked up & closed",
    .continueJourney: "Journey continued",
]
